### ISSUE TITLE
(see readme for the workaround) Cursor doesn't keep up with typing on some Samsung Android phones

### ISSUE THREAD

**Issue by richardhills** [catamphetamine/react-phone-number-input#75]
This reproducible on my Huawei running Android 6.0, and on Samsung S8+ phones. It happens with and without `type="tel"` in the `<Phone>` declaration, so I don't think it's related to [this issue](https://github.com/catamphetamine/react-phone-number-input/issues/59).

Set the country to "UK", and type in "077915". The input will contain "07791 5" (with a space) and the cursor will be positioned before the 5. Typing "67890" after this results in "07791 678905" instead of "07791 567890".

It appears that on these devices that doing a complete onChange -> setState -> render cycle results in the cursor being positioned incorrectly. Calling setState in a setTimeout callback seems to fix it.

Here's a pull request that seems to resolve the issue on my device: https://github.com/catamphetamine/react-phone-number-input/pull/74

**Comment by catamphetamine** [comment]
Did you install the latest version by the way?

**Comment by catamphetamine** [comment]
You can check that in the browser: if you open DevTools and select the `<input/>` it should not have `type="tel"`.
Then you can test it on the phone.

**Comment by Gedzis** [comment]
I have same issue with latest version and Samsung device. 
If I type LT phone "+370 6|" cursor is moved before 6, so it looks like "+370 |6"

**Comment by catamphetamine** [comment]
@Gedzis is `type="tel"` absent from the `<input/>` tag on your webpage?

**Comment by catamphetamine** [comment]
The pull request relies on `setTimeout` which is magic and I'm not willing to magically fix things preferring durable and clear solutions instead.
The UK use case is too simple to verify the workaround too because it's the simplest one and is most likely a false positive:

Use case:

Select `UK` country

```
0|
07|
077|
0779|
07791|
07791 |5
07791 6|5
07791 67|5
07791 678|5
07791 6789|5
07791 67890|5
```

So, from the use case it seems like the same issue as this one:
https://github.com/catamphetamine/input-format/issues/2
It seems like the caret position isn't controllable at all.

When a timeout is introduced the `<input/>` value goes out of sync and React just resets the caret positioning it in the end of the `<input/>` value.
This results in falsy confidence of the bug being "fixed".
Try that trick on the complex ones, having lots of braces and spaces in between, inserting digits somewhere in the middle.

**Comment by Gedzis** [comment]
@catamphetamine I agree that setting timeout isn't best solution.
Issue that I am having: https://ibb.co/cNtbWv

**Comment by richardhills** [comment]
Hi @catamphetamine , yes it's reproducible on the latest version. It happens with and without `type="tel"` in the `<Phone>` declaration.

I'm just trying to identify exactly when it goes wrong in the onChange -> render cycle.

**Comment by catamphetamine** [comment]
If you want my directions then provided this simulation it's the caret positioning code:

```
0|
07|
077|
0779|
07791|
07791 |5
07791 6|5
07791 67|5
07791 678|5
07791 6789|5
07791 67890|5
```

i.e. the caret always moves 1 step forward and is uncontrollable through `input-format`'s `setCaretPosition()` for some reason.
ppl said "Google's keyboard" worked whatever that means.

**Comment by richardhills** [comment]
Hi @catamphetamine, I've realised it's more complex. I get:

```
0|
07|
077|
0779|
07791|
07791 |5
0779165|
07791 65|7
07791 658|7
07791 6589|7
07791 65890|7
```

I've stepped through the code, and `setCaretPosition()` is called correctly. At some point after `render()` is called, the caret is placed in the incorrect position.

**Comment by richardhills** [comment]
Also, the cursor is updated correctly by setCaretPosition(), but it reverts to the incorrect place.

**Comment by catamphetamine** [comment]
My feeling is like on Androids the keyboard is stubborn and forces itself through any programmatic cursor positioning.
Not sure if something like `setCaretPosition(input, 123)` would work when hooked to a button tap or executed in a timeout.
If it does, then a manual `input.addEventListener('change', () => setCaretPosition(...))` could be tried, i.e. without React at all, to see if it allows messing with caret position at all inside the 'change' event handler.
Guesses all around.

**Comment by catamphetamine** [comment]
And if it was React resetting the caret position then AFAIK it always resets to the end, which is not the case in your case.

**Comment by richardhills** [comment]
Calling `setCaretPosition(input, 123)` works as expected. The keyboard appears to respect the cursor position, but it's being reverted afterwards. I'm not sure why react would necessarily reset it to the end, but I'll continue investigating.

**Comment by catamphetamine** [comment]
So, your point is that the cursor position only resets in the React case and that in a simple event hook `setCaretPosition()` results aren't reverted?
React is known to reset caret position to the end when `input.value` changes in any way. It's the reason `input-format` manually does `input.value = xxx` before calling `onChange(xxx)`.

**Comment by richardhills** [comment]
@catamphetamine , that's right. I'm trying to breakpoint in places where react might update the carot position with no luck.

**Comment by catamphetamine** [comment]
Yeah, that's a tricky one.
I imagine it could take up to infinity time to debug this one so make sure your head doesn't blow ha ha.

**Comment by shivgarg5676** [comment]
Is there any update on this bug. I am facing same problem on ios as well

**Comment by catamphetamine** [comment]
@shivgarg5676 This bug is for Android only.
If you're having a bug on iOS then open a separate issue and post reproduction instructions.

**Comment by shivgarg5676** [comment]
Yes, this is on android only and that too some limited devices. But are we able to address this

**Comment by catamphetamine** [comment]
@shivgarg5676 My guess is that this bug is unbeatable.

**Comment by shivgarg5676** [comment]
Is there any way we can disable this spacing and formatting feature. This will automatically disable this bug.

**Comment by catamphetamine** [comment]
@shivgarg5676 Use `libphonenumber-js` `parse()` if you don't want the formatting feature.

**Comment by jcollum** [comment]
We're seeing it too. It happens on Samsung 7.0 devices but does not happen on: 

- LG 7.0 device

- AVD emulator 

- Chrome emulating Galaxy S5

**Comment by catamphetamine** [comment]
@jcollum I know on iPhones there's a debugging feature: one can hook up an iPhone to a macOS machine and then observe any `console.log()`s or even use the DevTools debugger.
Is there something like that on Android phones?

**Comment by jcollum** [comment]
Yes, pretty sure you can do that with Chrome debugging tools: `chrome://inspect/#devices`. I'll post more here when I get a real Samsung device to test with.

**Comment by catamphetamine** [comment]
For anyone interested, copy-pasting from the PR: https://github.com/catamphetamine/react-phone-number-input/pull/74#issuecomment-372338179

> If anyone wants they can send a phone to my US mailbox this week (a window of opportunity) so that I could take a look at the issue when the package arrives from US to Russia (I'm purchasing something this week).

**Comment by Xopsy** [comment]
Hi @catamphetamine , yes you can debug on physical mobile phone in android. You can use chrome, firefox, even with safari. After you give a your phone developer permissions and you can set up debugging with usb, you can use that features.

Firefox --> Firefox WebIDE
Chrome --> chrome://inspect/#devices

Even if you can use adb or android studio for that purpose and into console view get different logs.

Only I leave this message because I saw your before answer. I didn't try yet this component, but I was reading different threads about this issues. I don't have to much time time but if possible I'd try to see this issue for myself.

But for my experience a delay something, it's not usually good solution. Because of ent lop work in javascript

https://developer.mozilla.org/en-US/docs/Web/JavaScript/EventLoop

If you need to do something after setstate is called, or you use callback param or lifecycle into react components like https://reactjs.org/docs/react-component.html#componentdidupdate to do something with DOM.

Thanks ;)

**Comment by brianjoseff** [comment]
Maybe if there's an elegant way to remove `AsYouType` as a dependency from input-control.js. I think the AsYouType formatter is the source of these issues (well Samsung, et al is...but anyway.)

Create like an "autoFormat" prop, and if false, then that flag is carried into input-control.js and it only passes the parsed input into the formatter if autoFormat flag is true (default)

**Comment by catamphetamine** [comment]
So, I implemented the feature.
See the linked commit for reference.

`react-phone-number-input@1.0.8`

Usage:

```js
import Phone, { BasicInput } from 'react-phone-number-input'

<Phone
  inputComponent={BasicInput}
  value={this.state.value}
  onChange={value => this.setState(value)}/>
```

**Comment by Xopsy** [comment]
Great fix @catamphetamine, I have been testing today from my device Samsung Galaxy S7 edge, everything is working fine. I tried to reproduce different issues commented above, and they work well, also I've tried other possibilities and it works fine.

**Comment by catamphetamine** [comment]
I also rewrote `<BasicInput/>`'s fix for React caret jumping today.
It was using direct `ReactDOM.findDOMNode(this.ref).value = event.target.value` in `1.0.8`.
In the new `react-phone-number-input@1.0.9` this `ReactDOM.findDOMNode` approach was replaced with storing `value` in `this.state`.
Because I don't have a Galaxy phone I don't know if `1.0.9` still works.
We'll see what different people say.

**Comment by Xopsy** [comment]
Hey @catamphetamine, I did test right now with new version and still working fine in my samsung phone.

I've tested to different browsers from Samsung Galaxy S7 Edge:

Chrome ✅ 
Firefox ✅ 
[Samsung Internet Browser](https://play.google.com/store/apps/details?id=com.sec.android.app.sbrowser&hl=en) ❌

Even I have simulated with Xcode iPhone X (Safari Browser ✅) is working fine.

But for main mobile browser (Chrome and Safari) is working fine.

The problem only persist with that Samsung browser but this one is a minor issue, I haven't tried other browsers. You have more [here](https://netmarketshare.com/browser-market-share.aspx?options=%7B%22filter%22%3A%7B%22%24and%22%3A%5B%7B%22deviceType%22%3A%7B%22%24in%22%3A%5B%22Mobile%22%5D%7D%7D%5D%7D%2C%22dateLabel%22%3A%22Trend%22%2C%22attributes%22%3A%22share%22%2C%22group%22%3A%22browser%22%2C%22sort%22%3A%7B%22share%22%3A-1%7D%2C%22id%22%3A%22browsersMobile%22%2C%22dateInterval%22%3A%22Monthly%22%2C%22dateStart%22%3A%222017-04%22%2C%22dateEnd%22%3A%222018-03%22%2C%22segments%22%3A%22-1000%22%7D).

**Comment by catamphetamine** [comment]
@Xopsy wow, thanks for thorough testing 👍 
Ok then, it works.

Samsung seems to a really big company that can afford writing its own everything: their own Android shell, their own Bixby, their own web browser... Looks like another Microsoft to me.

**Comment by catamphetamine** [comment]
Update: I released version `2.x` which defaults to `smartCaret={false}`.

**Comment by catamphetamine** [comment]
Since `smartCaret` is now `false` by default I'll close this issue until someone reports otherwise.
Btw, I released version `2.1` which now only supports React >= 16.3

**Comment by catamphetamine** [comment]
FYI: released version `react-phone-number-input@2.5.0` where `smartCaret` is `true` by default again. The reason is that I searched for popular mask input libraries and found [`text-mask`](https://github.com/text-mask/text-mask) to be one of the most popular. I searched its issues and found the exact same one. And I just copied their [solution](https://github.com/catamphetamine/input-format/issues/2#issuecomment-552032606): wrapping `onChange` in `setTimeout(0)`. We'll see how users report if it works on weird Androids now.
Updated the [demo](http://catamphetamine.github.io/react-phone-number-input/).
